tests/phpunit/api/v4/Entity/QueueTest.php

```diff
@@ -0,0 +1,409 @@
+      'type' => 'Sql',
+      'runner' => 'task',
+      'error' => 'delete',
+      'retry_limit' => 2,
+      'retry_interval' => 0,
+      'lease_time' => 1,
+    ]);
+    $this->assertEquals(0, $queue->numberOfItems());
+
+    \Civi::queue($queueName)->createItem(new \CRM_Queue_Task(
+      [QueueTest::class, 'doSomething'],
+      ['playinghooky']
+    ));
+    $this->assertEquals(1, $queue->numberOfItems());
+
+    $claim1 = $this->waitForClaim(0.5, 5, $queueName);
+    // Oops, don't do anything with claim #1!
+    $this->assertEquals(1, $queue->numberOfItems());
+    $this->assertEquals([], \Civi::$statics[__CLASS__]['doSomethingLog']);
+
+    $claim2 = $this->waitForClaim(0.5, 5, $queueName);
+    // Oops, don't do anything with claim #2!
+    $this->assertEquals(1, $queue->numberOfItems());
+    $this->assertEquals([], \Civi::$statics[__CLASS__]['doSomethingLog']);
+
+    $claim3 = $this->waitForClaim(0.5, 5, $queueName);
+    $this->assertEquals(1, $queue->numberOfItems());
+    $result = Queue::runItems(0)->setItems([$claim3])->execute()->first();
+    $this->assertEquals(0, $queue->numberOfItems());
+    $this->assertEquals(['playinghooky_ok'], \Civi::$statics[__CLASS__]['doSomethingLog']);
+    $this->assertEquals('ok', $result['outcome']);
+  }
+
+  /**
+   * Add a task which is never going to succeed. The task fails every time, and eventually
+   * we either delete it or abort the queue.
+   *
+   * @param string $errorMode
+   *   Either 'delete' or 'abort'
+   * @dataProvider getErrorModes
+   */
+  public function testRetryWithEventualFailure(string $errorMode) {
+    \Civi::$statics[__CLASS__]['doSomethingResult'] = FALSE;
+
+    $queueName = 'QueueTest_' . md5(random_bytes(32)) . '_linear';
+    $queue = \Civi::queue($queueName, [
+      'type' => 'Sql',
+      'runner' => 'task',
+      'error' => $errorMode,
+      'retry_limit' => 2,
+      'retry_interval' => 0,
+      'lease_time' => 1,
+    ]);
+    $this->assertEquals(0, $queue->numberOfItems());
+
+    \Civi::queue($queueName)->createItem(new \CRM_Queue_Task(
+      [QueueTest::class, 'doSomething'],
+      ['playinghooky']
+    ));
+    $this->assertEquals(1, $queue->numberOfItems());
+
+    $claimAndRun = function($expectOutcome, $expectEndCount) use ($queue, $queueName) {
+      $claim = $this->waitForClaim(0.5, 5, $queueName);
+      $this->assertEquals(1, $queue->numberOfItems());
+      $result = Queue::runItems(0)->setItems([$claim])->execute()->first();
+      $this->assertEquals($expectEndCount, $queue->numberOfItems());
+      $this->assertEquals($expectOutcome, $result['outcome']);
+    };
+
+    $claimAndRun('retry', 1);
+    $claimAndRun('retry', 1);
+    switch ($errorMode) {
+      case 'delete':
+        $claimAndRun('delete', 0);
+        $this->assertEquals(TRUE, $queue->isActive());
+        break;
+
+      case 'abort':
+        $claimAndRun('abort', 1);
+        $this->assertEquals(FALSE, $queue->isActive());
+        break;
+    }
```

Comment:
@eileenmcnaughton This bit shows how the `error=>delete` and `error=>abort` lead to different outcomes (after a series of retries).